### PR TITLE
Attach Path Condition to Optimized Functions

### PR DESCRIPTION
**Pull request by cblappert**
Release Notes: None

Optimized functions will now be evaluated with the path condition at the time of function closure creation.

Resolves issue #2422

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -197,5 +197,13 @@ export function createModelledFunctionCall(
     thisValue !== undefined
       ? thisValue
       : AbstractValue.createAbstractArgument(realm, "this", funcValue.expressionLocation, ObjectValue);
-  return () => call(thisArg, args);
+  return () => {
+    let savedPathConditions = realm.pathConditions;
+    let newPathConditions = funcValue.pathConditionDuringDeclaration;
+    invariant(newPathConditions);
+    realm.pathConditions = newPathConditions;
```
Comment (NTillmann) [inline src/utils.js]: Let's wrap this in a try-finally.

File: src/utils/paths.js [hunk 2]
```diff
@@ -112,6 +112,16 @@ export class PathConditionsImplementation extends PathConditions {
       savedBaseConditions.refineBaseConditons(realm);
     }
   }
+
+  clone(): PathConditionsImplementation {
```
Comment (NTillmann) [inline src/utils/paths.js]: @hermanventer says we shouldn't have to clone, is that right?

Comment (NTillmann) [reply src/utils/paths.js]: Maybe not clone, but wrap the realm path conditions in a new path conditions object. In any case, watch this PR: #2538

File: src/methods/function.js [hunk 3]
```diff
@@ -1252,7 +1252,10 @@ export class FunctionImplementation {
     }
 
     // 5. Return FunctionInitialize(F, kind, ParameterList, Body, Scope).
-    return this.FunctionInitialize(realm, F, kind, ParameterList, Body, Scope);
+    let result = this.FunctionInitialize(realm, F, kind, ParameterList, Body, Scope);
+    invariant(F.pathConditionDuringDeclaration === undefined, "Function should only have one declaration site");
+    F.pathConditionDuringDeclaration = realm.pathConditions.clone();
```
Comment (hermanventer) [inline src/methods/function.js]: Do not clone.

File: src/types.js [hunk 4]
```diff
@@ -386,6 +386,10 @@ export class PathConditions {
   }
 
   refineBaseConditons(realm: Realm, depth?: number = 0): void {}
+
+  clone(): PathConditions {
```
Comment (hermanventer) [inline src/types.js]: Remove this.

File: src/utils/paths.js (hunk 2 above)
Comment (hermanventer) [reply src/utils/paths.js]: Do not clone.

File: src/methods/function.js [hunk 5]
```diff
@@ -1254,7 +1254,7 @@ export class FunctionImplementation {
     // 5. Return FunctionInitialize(F, kind, ParameterList, Body, Scope).
     let result = this.FunctionInitialize(realm, F, kind, ParameterList, Body, Scope);
     invariant(F.pathConditionDuringDeclaration === undefined, "Function should only have one declaration site");
-    F.pathConditionDuringDeclaration = realm.pathConditions.clone();
+    F.pathConditionDuringDeclaration = realm.pathConditions;
```
Comment (NTillmann) [inline src/methods/function.js]: Since we don't clone anymore, add
```js
invariant(realm.pathConditions.isReadOnly);
```

**Pull request closed without merging**